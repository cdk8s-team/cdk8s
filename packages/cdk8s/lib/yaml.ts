import * as fs from 'fs';
import * as YAML from 'yaml';
import { execFileSync } from 'child_process';
import * as os from 'os';
import * as path from 'path';

/**
 * YAML utilities.
 */
export class Yaml {
  /**
   * Saves a set of objects as a multi-document YAML file.
   * @param filePath The output path
   * @param docs The set of objects
   */
  public static save(filePath: string, docs: any[]) {
    // convert each resource to yaml and separate with a '---' line
    const data = docs.map(r => YAML.stringify(r)).join('---\n');
    fs.writeFileSync(filePath, data, { encoding: 'utf-8' });
  }

  /**
   * Saves a set of YAML documents into a temp file (in /tmp)
   *
   * @returns the path to the temporary file
   * @param docs the set of documents to save
   */
  public static tmp(docs: any[]): string {
    const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s-'));
    const filePath = path.join(tmpdir, 'temp.yaml');
    Yaml.save(filePath, docs);
    return filePath;
  }

  /**
   * Downloads a set of YAML documents (k8s manifest for example) from a URL or
   * a file and returns them as javascript objects.
   *
   * Empty documents are filtered out.
   *
   * @param urlOrFile a URL of a file path to load from
   * @returns an array of objects, each represents a document inside the YAML
   */
  public static load(urlOrFile: string): any[] {
    const body = loadurl(urlOrFile);
    const objects = YAML.parseAllDocuments(body);
    const result = new Array<any>();

    for (const obj of objects.map(x => x.toJSON())) {
      // skip empty documents
      if (obj === undefined) { continue; }
      if (obj === null) { continue; }
      if (Array.isArray(obj) && obj.length === 0) { continue; }
      if (typeof(obj) === 'object' && Object.keys(obj).length === 0) { continue; }

      result.push(obj);
    }

    return result;
  }

  /**
   * Utility class.
   */
  private constructor() {
    return;
  }
}

/**
 * Loads a url (or file) and returns the contents.
 * This method spawns a child process in order to perform an http call synchronously.
 */
function loadurl(url: string): string {
  const loadurl = path.join(__dirname, '_loadurl.js');
  return execFileSync(process.execPath, [ loadurl, url ], { encoding: 'utf-8' }).toString()
}

