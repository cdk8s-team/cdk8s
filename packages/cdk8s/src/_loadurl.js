// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

const { http, https } = require('follow-redirects');
const { parse } = require('url');
const fs = require('fs');

const url = process.argv[2];
if (!url) {
  console.error(`Usage: ${process.argv[1]} URL`);
  process.exit(1);
}

const purl = parse(url);

if (!purl.protocol) {
  if (!purl.pathname) {
    throw new Error(`unable to parse pathname from file url: ${url}`);
  }

  process.stdout.write(fs.readFileSync(purl.pathname, 'utf-8'));
  process.exit(0);
}

const client = getHttpClient();

const get = client.get(url, response => {
  if (response.statusCode !== 200) {
    throw new Error(`${response.statusCode} response from http get: ${response.statusMessage}`);
  }

  response.on('data', chunk => process.stdout.write(chunk));
});

get.once('error', err => {
  throw new Error(`http error: ${err.message}`);
});

function getHttpClient() {
  switch (purl.protocol) {
    case 'http:': return http;
    case 'https:': return https;
    default:
      throw new Error(`unsupported protocol "${purl.protocol}" in url: ${url}`);
  }
}
