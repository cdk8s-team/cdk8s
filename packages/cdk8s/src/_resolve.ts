import { Lazy } from './lazy';

export function resolve(value: any): any {

  if (value == null) {
    return value;
  }

  // cdk8s token
  if (value instanceof Lazy) {
    const resolved = value.produce();
    return resolve(resolved);
  }

  // implicit token as generated by the cdk8s-cli (e.g IntOrString)
  if (typeof(value.resolve) === 'function') {
    const resolved = value.resolve();
    return resolve(resolved);
  }

  if (typeof(value) !== 'object') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(x => resolve(x));
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    result[k] = resolve(v);
  }

  return result;

}
