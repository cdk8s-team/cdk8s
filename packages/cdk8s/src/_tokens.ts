import { Tokenization, DefaultTokenResolver, StringConcat, Construct } from 'constructs';

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

export function resolve<T>(scope: Construct, obj: T): T {
  return Tokenization.resolve(obj, {
    scope,
    resolver: TOKEN_RESOLVER,
  });
}