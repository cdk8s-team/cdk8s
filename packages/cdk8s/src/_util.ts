export interface SanitizeOptions {
  /**
   * Do not include empty objects (no keys).
   */
  readonly filterEmptyObjects?: boolean;

  /**
   * Do not include arrays with no items.
   */
  readonly filterEmptyArrays?: boolean;
}

export function sanitizeValue(obj: any, options: SanitizeOptions = { }): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {

    if (options.filterEmptyArrays && obj.length === 0) {
      return undefined;
    }

    return obj.map(x => sanitizeValue(x, options));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = sanitizeValue(value, options);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  if (options.filterEmptyObjects && Object.keys(newObj).length === 0) {
    return undefined;
  }

  return newObj;
}
