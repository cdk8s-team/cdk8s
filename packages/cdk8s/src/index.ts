export * from './api-object';
export * from './chart';
export * from './testing';
export * from './app';
export * from './include';
export * from './yaml';
export * from './metadata';