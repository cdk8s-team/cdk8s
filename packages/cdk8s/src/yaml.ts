import { execFileSync } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as YAML from 'yaml';
import { Type } from 'yaml/util';

// Ensure that all strings are quoted when written to yaml to avoid unquoted
// primitive types in the output yaml in fields that require strings.
YAML.scalarOptions.str.defaultType = Type.QUOTE_DOUBLE;

/**
 * YAML utilities.
 */
export class Yaml {
  /**
   * Saves a set of objects as a multi-document YAML file.
   * @param filePath The output path
   * @param docs The set of objects
   */
  public static save(filePath: string, docs: any[]) {
    // convert each resource to yaml and separate with a '---' line
    // NOTE: we convert undefined values to null, but ignore any documents that
    //  are undefined
    const data = docs.map(
      r => r === undefined ? '\n' : YAML.stringify(r, { keepUndefined: true }),
    ).join('---\n');
    fs.writeFileSync(filePath, data, { encoding: 'utf-8' });
  }

  /**
   * Saves a set of YAML documents into a temp file (in /tmp)
   *
   * @returns the path to the temporary file
   * @param docs the set of documents to save
   */
  public static tmp(docs: any[]): string {
    const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s-'));
    const filePath = path.join(tmpdir, 'temp.yaml');
    Yaml.save(filePath, docs);
    return filePath;
  }

  /**
   * Downloads a set of YAML documents (k8s manifest for example) from a URL or
   * a file and returns them as javascript objects.
   *
   * Empty documents are filtered out.
   *
   * @param urlOrFile a URL of a file path to load from
   * @returns an array of objects, each represents a document inside the YAML
   */
  public static load(urlOrFile: string): any[] {
    const body = loadurl(urlOrFile);

    // we need to use yaml-1.1 so that octal numbers in the form `0775` will be parsed
    // correctly (see https://github.com/eemeli/yaml/issues/205)
    const objects = YAML.parseAllDocuments(body, { schema: 'yaml-1.1' });
    const result = new Array<any>();

    for (const obj of objects.map(x => x.toJSON())) {
      // skip empty documents
      if (obj === undefined) { continue; }
      if (obj === null) { continue; }
      if (Array.isArray(obj) && obj.length === 0) { continue; }
      if (typeof(obj) === 'object' && Object.keys(obj).length === 0) { continue; }

      result.push(obj);
    }

    return result;
  }

  /**
   * Utility class.
   */
  private constructor() {
    return;
  }
}

/**
 * Loads a url (or file) and returns the contents.
 * This method spawns a child process in order to perform an http call synchronously.
 */
function loadurl(url: string): string {
  const script = path.join(__dirname, '_loadurl.js');
  return execFileSync(process.execPath, [script, url], { encoding: 'utf-8' }).toString();
}
