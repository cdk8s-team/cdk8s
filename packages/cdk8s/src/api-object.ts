import { Construct, IConstruct, Node } from 'constructs';
import { Chart } from './chart';
import { sanitizeValue } from './_util';
import { resolve } from './_tokens';
import * as stringify from 'json-stable-stringify';
import { ApiObjectMetadata, ApiObjectMetadataDefinition } from './metadata';

/**
 * Options for defining API objects.
 */
export interface ApiObjectOptions {
  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {
  /**
   * The name of the API object.
   *
   * If a name is specified in `metadata.name` this will be the name returned.
   * Otherwise, a name will be generated by calling
   * `Chart.of(this).generatedObjectName(this)`, which by default uses the
   * construct path to generate a DNS-compatible name for the resource.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  /**
   * The chart in which this object is defined.
   */
  public readonly chart: Chart;

  /**
   * Metadata associated with this API object.
   */
  public readonly metadata: ApiObjectMetadataDefinition;

  /**
   * Defines an API object.
   *
   * @param scope the construct scope
   * @param ns namespace
   * @param options options
   */
  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.chart = Chart.of(this);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;

    this.name = options.metadata?.name ?? this.chart.generateObjectName(this);

    this.metadata = new ApiObjectMetadataDefinition({
      name: this.name,
      namespace: options.metadata?.namespace ?? this.chart.namespace,

      // override user defined values
      ...options.metadata,
    });
  }

  /**
   * Create a dependency between this ApiObject and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const data = {
      ...this.options,
      metadata: this.metadata.toJson(),
    };

    // convert to "pure data" so, for example, when we convert to yaml these
    // references are not converted to anchors.
    return JSON.parse(stringify(sanitizeValue(resolve(this, data))));
  }
}
