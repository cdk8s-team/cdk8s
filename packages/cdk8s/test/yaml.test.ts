import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { Yaml } from '../src/yaml';

describe('load', () => {
  test('from file', () => {
    expect(Yaml.load(`${__dirname}/fixtures/sample.yaml`)).toMatchSnapshot();
  });

  test('from url', () => {
    expect(Yaml.load('https://raw.githubusercontent.com/kubernetes/examples/52158f68ddc508eac9f15a9bb7260b838e644c9f/guestbook/all-in-one/guestbook-all-in-one.yaml')).toMatchSnapshot();
  });

  test('empty documents are filtered out', () => {
    const file = Yaml.tmp([
      { doc: 1 },
      null, // filtered
      'str_doc', // not filtered
      { }, // filtered
      '', // not filtered
      undefined, // filtered
      0, // not filtered
      [], // filtered
      { doc: 2 },
    ]);

    expect(Yaml.load(file)).toMatchSnapshot();
  });
});

describe('save', () => {
  test('single document', () => {
    const outputFile = Yaml.tmp([{ foo: 'bar', hello: [1, 2, 3] }]);
    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot();
  });

  test('multiple documents', () => {
    const outputFile = Yaml.tmp([
      { foo: 'bar', hello: [1, 2, 3] },
      { number: 2 },
    ]);

    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot();
  });

  test('empty values are preserved', () => {
    const temp = Yaml.tmp([
      {
        i_am_undefined: undefined, // converted to "null"
        i_am_null: null,
        empty_array: [],
        empty_object: {},
      },
    ]);

    expect(fs.readFileSync(temp, 'utf-8')).toMatchSnapshot();
  });

  test('empty documents are respected', () => {
    const outputFile = Yaml.tmp([
      {},
      {},
      undefined,
      { empty: true },
      {},
    ]);

    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot();
  });

  test('strings are quoted', () => {
    const outputFile = Yaml.tmp([{
      foo: 'on',
      bar: 'this has a "big quote"',
      not_a_string: true,
    }]);

    expect(fs.readFileSync(outputFile, 'utf-8')).toMatchSnapshot();
  });

});

test('yaml 1.1 octal numbers are parsed correctly', () => {
  const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'cdk8s-'));
  const filePath = path.join(tmpdir, 'temp.yaml');
  fs.writeFileSync(filePath, 'foo: 0755');

  expect(Yaml.load(filePath)).toEqual([{ foo: 493 }]);
});
