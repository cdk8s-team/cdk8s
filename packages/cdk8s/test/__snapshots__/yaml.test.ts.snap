// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load empty documents are filtered out 1`] = `
Array [
  Object {
    "doc": 1,
  },
  "str_doc",
  "",
  0,
  Object {
    "doc": 2,
  },
]
`;

exports[`load from file 1`] = `
Array [
  Object {
    "hello": 1234,
    "world": 111,
  },
  Object {
    "foo": Array [
      "bar",
      "zoo",
      "goo",
    ],
  },
  Array [
    "hello",
    "world",
  ],
]
`;

exports[`load from url 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "redis",
        "role": "master",
        "tier": "backend",
      },
      "name": "redis-master",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
          "targetPort": 6379,
        },
      ],
      "selector": Object {
        "app": "redis",
        "role": "master",
        "tier": "backend",
      },
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "redis-master",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "redis",
          "role": "master",
          "tier": "backend",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "redis",
            "role": "master",
            "tier": "backend",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "k8s.gcr.io/redis:e2e",
              "name": "master",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "redis",
        "role": "slave",
        "tier": "backend",
      },
      "name": "redis-slave",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 6379,
        },
      ],
      "selector": Object {
        "app": "redis",
        "role": "slave",
        "tier": "backend",
      },
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "redis-slave",
    },
    "spec": Object {
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "app": "redis",
          "role": "slave",
          "tier": "backend",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "redis",
            "role": "slave",
            "tier": "backend",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GET_HOSTS_FROM",
                  "value": "dns",
                },
              ],
              "image": "gcr.io/google_samples/gb-redisslave:v1",
              "name": "slave",
              "ports": Array [
                Object {
                  "containerPort": 6379,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app": "guestbook",
        "tier": "frontend",
      },
      "name": "frontend",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
        },
      ],
      "selector": Object {
        "app": "guestbook",
        "tier": "frontend",
      },
      "type": "NodePort",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "frontend",
    },
    "spec": Object {
      "replicas": 3,
      "selector": Object {
        "matchLabels": Object {
          "app": "guestbook",
          "tier": "frontend",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "guestbook",
            "tier": "frontend",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "GET_HOSTS_FROM",
                  "value": "dns",
                },
              ],
              "image": "gcr.io/google-samples/gb-frontend:v4",
              "name": "php-redis",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
              "resources": Object {
                "requests": Object {
                  "cpu": "100m",
                  "memory": "100Mi",
                },
              },
            },
          ],
        },
      },
    },
  },
]
`;

exports[`save empty documents are respected 1`] = `
"{}
---
{}
---

---
empty: true
---
{}
"
`;

exports[`save empty values are preserved 1`] = `
"i_am_undefined: null
i_am_null: null
empty_array: []
empty_object: {}
"
`;

exports[`save multiple documents 1`] = `
"foo: bar
hello:
  - 1
  - 2
  - 3
---
number: 2
"
`;

exports[`save single document 1`] = `
"foo: bar
hello:
  - 1
  - 2
  - 3
"
`;
